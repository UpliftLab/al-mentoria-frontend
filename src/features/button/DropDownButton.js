import PropTypes from 'prop-types';
import CaretDown from '../../images/caret-down.svg';

const DropDownButton = ({
  options,
  elementID,
  defaultOption,
  required,
  onChange,
}) => (
  <select
    id={elementID}
    defaultValue=""
    required={required}
    onChange={onChange}
    className="px-6 py-2 rounded-full font-semibold min-w-[200px] bg-lime-500 text-white border-2 border-white focus:outline-none appearance-none"
    style={{
      backgroundImage: `url(${CaretDown})`,
      backgroundSize: '1.2rem',
      backgroundPosition: 'right 1rem center',
      backgroundRepeat: 'no-repeat',
    }}
  >
    <option value="" disabled>{defaultOption}</option>
    {options.map((value) => <option key={value.id} value={value.id}>{value.text}</option>)}
  </select>
);

DropDownButton.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
  })).isRequired,
  elementID: PropTypes.string.isRequired,
  defaultOption: PropTypes.string.isRequired,
  required: PropTypes.bool,
  onChange: PropTypes.func,
};

DropDownButton.defaultProps = {
  required: false,
  onChange: () => {},
};

export default DropDownButton;
