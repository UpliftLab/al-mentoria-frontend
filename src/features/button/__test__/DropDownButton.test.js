import {
  render,
  screen,
  cleanup,
  fireEvent,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import DropDownButton from '../DropDownButton';

describe('DropDownButton', () => {
  let options;
  let defaultOption;
  let elementID;

  beforeEach(() => {
    const arr = new Array(10).fill(0);
    options = arr.map((_, i) => ({ id: i, text: `text ${i}` }));
    defaultOption = 'Select a Topic';
    elementID = 'DropDownID';
  });

  afterEach(cleanup);

  it('renders correctly in the DOM', () => {
    const tree = renderer
      .create(<DropDownButton
        options={options}
        defaultOption={defaultOption}
        elementID={elementID}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with the green background and white text', () => {
    render(
      <DropDownButton
        options={options}
        defaultOption={defaultOption}
        elementID={elementID}
      />,
    );

    const select = screen.getByRole('combobox');
    const { classList } = select;

    expect(classList).toContain('bg-lime-500');
    expect(classList).toContain('text-white');
  });

  it('renders with the default option', () => {
    render(
      <DropDownButton
        options={options}
        defaultOption={defaultOption}
        elementID={elementID}
      />,
    );

    const select = screen.getByDisplayValue(defaultOption);

    expect(select).toBeInTheDocument();
  });

  it('permits the selection of an option', () => {
    const { container } = render(
      <DropDownButton
        options={options}
        defaultOption={defaultOption}
        elementID={elementID}
      />,
    );

    const select = container.querySelector(`#${elementID}`);

    fireEvent.change(select, { target: { value: 2 } });

    expect(select.selectedIndex).toBe(select.options[2].index + 1);
  });
});
