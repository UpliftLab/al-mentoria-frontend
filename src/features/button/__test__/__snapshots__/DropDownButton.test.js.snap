// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownButton renders correctly in the DOM 1`] = `
<select
  className="px-6 py-2 rounded-full font-semibold min-w-[200px] bg-lime-500 text-white border-2 border-white focus:outline-none appearance-none"
  defaultValue=""
  id="DropDownID"
  onChange={[Function]}
  required={false}
  style={
    Object {
      "backgroundImage": "url(caret-down.svg)",
      "backgroundPosition": "right 1rem center",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "1.2rem",
    }
  }
>
  <option
    disabled={true}
    value=""
  >
    Select a Topic
  </option>
  <option
    value={0}
  >
    text 0
  </option>
  <option
    value={1}
  >
    text 1
  </option>
  <option
    value={2}
  >
    text 2
  </option>
  <option
    value={3}
  >
    text 3
  </option>
  <option
    value={4}
  >
    text 4
  </option>
  <option
    value={5}
  >
    text 5
  </option>
  <option
    value={6}
  >
    text 6
  </option>
  <option
    value={7}
  >
    text 7
  </option>
  <option
    value={8}
  >
    text 8
  </option>
  <option
    value={9}
  >
    text 9
  </option>
</select>
`;
